enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+set.
+
+If the administrator wants to disable automatic detection of the upgrade bundle,
+and wants to manually copy the file to one of the cluster nodes she will do
+something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /root/upgrade-4.13.7-x86_64.tar
+```
+
+If the administrator wants to disable automatic detection, but wants to use a
+USB stick anyhow he will do something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /dev/sdb
+```
+
+When the `status.desired.bundle.file` field is populated the cluster version
+operator will start to replicate the bundle to the rest of the nodes of the
+cluster. To do so it will first disable auto-scaling to ensure that no new
+nodes are added to the cluster while this process is in progress. Then it will
+start a new `bundle-server` daemon set. Each of the pods in this daemon set
+will check if the file specified in the `status.desired.bundle.file` field
+exists and contains a valid bundle. If it does then it will serve it via HTTP
+for the other nodes of the cluster. If it doesn't exist then it will do
+nothing.
+
+Simultaneously with the `bundle-server` daemon set the cluster version operator
+will also start a new `bundle-extractor` batch job in each node of the cluster.
+Each pod in these jobs will try to read the bundle from the location specified
+in `status.desired.bundle.file` field. If that file doesn't exist it will try
+to download it from the HTTP server of one of the pods of the `bundle-server`
+daemon set; the first that responds with `HTTP 200 Ok`. Once it has either the
+file or body of the HTTP response it will extract the contents of the
+`metadata.json` file (will always be the first entry of the tarball) to check
+that there is space in `/var/lib/upgrade` for the size indicated in the `size`
+field. If there is not enough space then it will be reported as en error
+condition in the `ClusterVersion` object and the upgrade process will be
+aborted.
+
+If the `spec.desiredUpdate.bundle.digest` field has been set set then the
+`bundle-extractor` will calculate the digest of the bundle, and if it doesn't
+match it will report it as an error condition in the `ClusterVersion` object and
+the upgrade process will be aborted.
+
+If there is enough space the `bundle-extractor` will extract the contents of
+the bundle to the `/var/lib/upgrade/4.13.7-x86_64` directory.
+
+Once the digest has been verified and the bundle has been completely extracted
+the `bundle-extractor` will update the status of the `ClusterVersion` object to
+indicate that the bundle is extracted in that node. This will be done via a new
+`status.desired.bundle.nodes.*.extracted` and
+`status.seried.bundle.nodes.*.metadata` fields. For example, when `node0` and
+`node2` have completed the extraction of the bundle but `node1` hasn't, the
+`ClusterVersion` status will look like this:
```

Comment:
I think there's some overlap here between CVO status of the daemonset and MCO status of the rollout of the pinned images.

If we can just have the CVO monitor rollout of the machineconfig including the pinned images, that sounds way simpler.

Reply:
I think this information is still needed. We need to know the image digests in order to write them to the `pinnedImages`  field of the  `ContainerRuntimeConfig`, to request pre-loading and pinning of the images. That is  before MCO gets into the picture.

We also need the `metadata.release` field in order to trigger the upgrade.

An alternative to this would be to teach CVO to inspect the release image and extract the payload digests from it directly. Maybe it is already doing it? But that would need to be done in two steps: first pre-load and pin the release image, and then run the equivalent of `oc adm release info` but in a way that doesn't need to contact the registry server.

Reply:
One problem with going through the CVO is that there are some *large* containers in the payload and we effectively rely on things only being fetched dynamically.